src/get_objects_on_table.py

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python3
+import open3d as o3d
+from open3d_ros_helper import open3d_ros_helper as orh
+import copy
+import rospy
+import numpy as np
+import tf2_ros
+from tf2_sensor_msgs.tf2_sensor_msgs import do_transform_cloud
+from scipy.spatial.transform import Rotation as R
+from vision_msgs.msg import BoundingBox3D, BoundingBox3DArray
+from table_plane_extractor.srv import GetObjectsOnTable, GetObjectsOnTableResponse
+from table_plane_extractor_srv import transformPointCloud, table_plane_extractor_methode
+
+target_frame = 'base_link'
+
+def get_objects_on_table(req):
+    '''
+    Service that returns bounding boxes of objects found on a table plane.
+    Input: sensor_msgs/PointCloud2 scene_pointcloud
+    Output: vision_msgs/BoundingBox3DArray detected_objects
+    '''
+    try:
+        # call directly as it's in the same package -> no ros overhead
+        response = table_plane_extractor_methode(req)
+    except Exception as e:
+        rospy.logerr(e)
+        raise rospy.ServiceException(e)
+
+    tf_buffer = tf2_ros.Buffer()
+    tf_listener = tf2_ros.TransformListener(tf_buffer)
+    scene_rel_to_camera = req.point_cloud
+    camera_frame_id = scene_rel_to_camera.header.frame_id
+
+
+    bb_arr = BoundingBox3DArray()
+    bb_arr.header.stamp = scene_rel_to_camera.header.stamp
+    bb_arr.header.frame_id = target_frame
+    bb_arr.boxes = []
+    for plane_cloud in response.clouds:
+        plane_o3d = orh.rospc_to_o3dpc(plane_cloud)
+        bb_cloud = plane_o3d.get_oriented_bounding_box()
+        # remove floor (plane is in map frame)
+        z_min = rospy.get_param('/get_objects_on_table/z_min')
+        if bb_cloud.center[2] < z_min:
+            continue
+        
+        # transform scene pointcloud to map frame
+        scene = transformPointCloud(scene_rel_to_camera, plane_cloud.header.frame_id, camera_frame_id, tf_buffer)
+        scene_o3d = orh.rospc_to_o3dpc(scene)
+
+        # get bounding box above table
+        bb_above_table = copy.deepcopy(bb_cloud)
+        bb_above_table.center = (bb_above_table.center[0], bb_above_table.center[1], bb_above_table.center[2]+0.2+bb_above_table.extent[2])
+        bb_above_table.extent = (bb_above_table.extent[0]+0.04, bb_above_table.extent[1]+0.04, bb_above_table.extent[2]+0.4)
```

Comment:
that's nice, but maybe this 0.4 parameter on the extent should be a parameter? It depends on the object size, and in the case of a shelf might be a bit tricky to set.

Reply:
Implemented